src/backend/distributed/commands/foreign_constraint.c

```diff
@@ -149,132 +155,110 @@ ErrorIfUnsupportedForeignConstraint(Relation relation, char distributionMethod,
 	while (HeapTupleIsValid(heapTuple))
 	{
 		Form_pg_constraint constraintForm = (Form_pg_constraint) GETSTRUCT(heapTuple);
-		bool singleReplicatedTable = true;
+		bool referencedIsDistributed = false;
+		char referencedDistMethod = 0;
+		Var *referencedDistKey = NULL;
+		bool referencingIsReferenceTable = false;
+		bool referencedIsReferenceTable = false;
+		int referencingAttrIndex = -1;
+		int referencedAttrIndex = -1;
+		bool referencingColumnsIncludeDistKey = false;
+		bool foreignConstraintOnDistKey = false;
 
 		if (constraintForm->contype != CONSTRAINT_FOREIGN)
 		{
 			heapTuple = systable_getnext(scanDescriptor);
 			continue;
 		}
 
-		/*
-		 * We should make this check in this loop because the error message will only
-		 * be given if the table has a foreign constraint and the table is a reference
-		 * table.
-		 */
-		if (distributionMethod == DISTRIBUTE_BY_NONE)
+		referencedTableId = constraintForm->confrelid;
+		selfReferencingTable = (referencingTableId == referencedTableId);
+
+		referencedIsDistributed = IsDistributedTable(referencedTableId);
+		if (!referencedIsDistributed && !selfReferencingTable)
 		{
-			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-							errmsg("cannot create foreign key constraint because "
-								   "reference tables are not supported as the "
-								   "referencing table of a foreign constraint"),
-							errdetail("Reference tables are only supported as the "
-									  "referenced table of a foreign key when the "
-									  "referencing table is a hash distributed "
-									  "table")));
+			ereport(ERROR, (errcode(ERRCODE_INVALID_TABLE_DEFINITION),
+							errmsg("cannot create foreign key constraint"),
+							errdetail("Referenced table must be a distributed table"
```

Comment:
maybe "a distributed table or a reference table"

Reply:
> maybe "a distributed table or a reference table"

` a reference table` part is in the next line